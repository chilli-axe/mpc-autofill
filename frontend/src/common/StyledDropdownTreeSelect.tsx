import DropdownTreeSelect from "react-dropdown-tree-select";
import styled from "styled-components";

export const StyledDropdownTreeSelect = styled(DropdownTreeSelect)`
  .tag {
    color: black;
    background-color: #dddddd;
  }
  .tag-remove {
    color: #666666;
  }

  .dropdown-trigger {
    border-radius: 0.25rem;
    background-color: white;
  }
  .dropdown-content {
    border-radius: 0.25rem;
  }

  .search {
    background-color: white;
    color: black;
  }
  .search::placeholder {
    color: black;
  }
  .placeholder {
    color: gray;
  }

  .toggle {
    font: normal normal normal 12px/1 bootstrap-icons;
    top: 2px;
    left: 2px;
  }

  .toggle.collapsed::after {
    content: "\F4FA";
  }

  .toggle.expanded::after {
    content: "\F2E6";
  }

  color: black;

  .root {
    padding: 0;
    margin: 0;
  }
`;
