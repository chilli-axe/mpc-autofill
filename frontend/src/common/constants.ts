import { CardType as CardTypeSchema } from "@/common/schema_types";
import { Cardstock, CardType, Faces, SortBy } from "@/common/types";
export const ProjectName = "MPC Autofill";
export const MakePlayingCards = "MakePlayingCards.com";
export const MakePlayingCardsURL = "https://www.makeplayingcards.com";

export const Card: CardType = CardTypeSchema.Card;
export const Cardback: CardType = CardTypeSchema.Cardback;
export const Token: CardType = CardTypeSchema.Token;

export const SelectedImageSeparator = "@";
export const CardTypeSeparator = ":";
export const FaceSeparator = "//";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";

export const ToggleButtonHeight = 38; // pixels
export const NavbarHeight = 50; // pixels - aligns with the natural height of the navbar
export const RibbonHeight = 54; // pixels
export const NavbarLogoHeight = 40; // pixels
export const ContentMaxWidth = 1200; // pixels - aligns with bootstrap's large breakpoint

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
  SearchResults = "searchResults",
  SampleCards = "sampleCards",
}

export const S27: Cardstock = "(S27) Smooth";
export const S30: Cardstock = "(S30) Standard Smooth";
export const S33: Cardstock = "(S33) Superior Smooth";
export const M31: Cardstock = "(M31) Linen";
export const P10: Cardstock = "(P10) Plastic";
export const Cardstocks: Array<Cardstock> = [S27, S30, S33, M31, P10];

export const CardstockFoilCompatibility: { [cardstock in Cardstock]: boolean } =
  {
    [S27]: true,
    [S30]: true,
    [S33]: true,
    [M31]: true,
    [P10]: false,
  };

export const GoogleDriveImageAPIURL =
  "https://script.google.com/macros/s/AKfycbw8laScKBfxda2Wb0g63gkYDBdy8NWNxINoC4xDOwnCQ3JMFdruam1MdmNmN4wI5k4/exec";

export const SearchResultsEndpointPageSize = 300;
export const CardEndpointPageSize = 1000;

export enum CSVHeaders {
  quantity = "Quantity",
  frontQuery = "Front",
  frontSelectedImage = "Front ID",
  backQuery = "Back",
  backSelectedImage = "Back ID",
}

export const ExploreDebounceMS = 700;
export const ExplorePageSize = 60;

export const SortByOptions: { [option in SortBy]: string } = {
  dateCreatedDescending: "Date Created (Newest-Oldest)",
  dateCreatedAscending: "Date Created (Oldest-Newest)",
  dateModifiedDescending: "Date Modified (Newest-Oldest)",
  dateModifiedAscending: "Date Modified (Oldest-Newest)",
  nameAscending: "Name (A-Z)",
  nameDescending: "Name (Z-A)",
};
