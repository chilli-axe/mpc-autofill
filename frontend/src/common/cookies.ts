/**
 * Data API for interacting with anything stored in cookies or local storage.
 */

// import Ajv2020 from "ajv/dist/2020";
import Cookies from "js-cookie";

import {
  BackendURLKey,
  CSRFKey,
  GoogleAnalyticsConsentKey,
  SearchSettingsKey,
} from "@/common/constants";
import { Convert } from "@/common/schema_types";
import { SearchSettings, SourceDocuments, SourceRow } from "@/common/types";
import { getSourceRowsFromSourceSettings } from "@/common/utils";
import { getDefaultSearchSettings } from "@/store/slices/searchSettingsSlice";

// import * as SearchSettingsSchema from "../../../schemas/schemas/SearchSettings.json";

// const ajv = new Ajv2020();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFKey);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

/**
 * Get search settings from localStorage data. If valid data is retrieved,
 * ensure that all `sourceDocuments` are included in the returned settings,
 * with any new sources that weren't previously included added to the end and enabled.
 */
export function getLocalStorageSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  const serialisedRawSettings = localStorage.getItem(SearchSettingsKey) ?? "{}";
  try {
    const searchSettings = Convert.toSearchSettings(serialisedRawSettings);
    // great, the user has valid search settings stored in their browser local storage.
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) =>
        parseInt(sourceDocument.pk)
      )
    );
    // types have to be narrowed here because quicktype doesn't support our SourceRow data structure :(
    const sources: Array<SourceRow> = getSourceRowsFromSourceSettings(
      searchSettings.sourceSettings
    );
    const sourceInLocalStorageSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    searchSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInLocalStorageSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return searchSettings;
  } catch (e) {
    // quicktype will throw an error if the user's stored settings do not match the schema
    // e.g. upon first page load
    // just return the default settings
    return getDefaultSearchSettings(sourceDocuments);
  }
}

export function setLocalStorageSearchSettings(settings: SearchSettings): void {
  localStorage.setItem(SearchSettingsKey, JSON.stringify(settings));
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentKey);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentKey, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getLocalStorageBackendURL() {
  return localStorage.getItem(BackendURLKey);
}

export function setLocalStorageBackendURL(url: string) {
  localStorage.setItem(BackendURLKey, url);
}

export function clearLocalStorageBackendURL(): void {
  localStorage.removeItem(BackendURLKey);
}

//# endregion
