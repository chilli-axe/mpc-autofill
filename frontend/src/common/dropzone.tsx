/**
 * A small wrapper around the `react-dropzone` Dropzone for consistent presentation.
 */

import React, { useCallback } from "react";
import { DropzoneRootProps, useDropzone } from "react-dropzone";
import styled from "styled-components";

import { useAppDispatch } from "@/common/types";
import { setNotification } from "@/store/slices/toastsSlice";

const getColor = (props: DropzoneRootProps) => {
  if (props.isDragAccept) {
    return "#00e676";
  }
  if (props.isDragReject) {
    return "#ff1744";
  }
  if (props.isFocused) {
    return "#2196f3";
  }
  return "#eeeeee";
};

const Container = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  border-width: 2px;
  border-radius: 2px;
  border-color: ${(props: DropzoneRootProps) => getColor(props)};
  border-style: dashed;
  color: #bdbdbd;
  outline: none;
  transition: border 0.24s ease-in-out;
  cursor: pointer;
`;

interface StyledDropzoneProps {
  mimeTypes: { [mimeType: string]: Array<string> };
  fileUploadCallback: {
    (fileContent: string | ArrayBuffer | null): void;
  };
  label: string;
  disabled: boolean;
}

export function TextFileDropzone({
  mimeTypes,
  fileUploadCallback,
  label,
  disabled,
}: StyledDropzoneProps) {
  const dispatch = useAppDispatch();

  // let the user know when something goes wrong with reading their file
  const onAbort = useCallback(
    () =>
      dispatch(
        setNotification([
          "dropzone-onabort",
          {
            name: "File upload error",
            message: "File reading was stopped.",
            level: "error",
          },
        ])
      ),
    [dispatch]
  );
  const onError = useCallback(
    () =>
      dispatch(
        setNotification([
          "dropzone-onerror",
          {
            name: "File upload error",
            message: "An error occurred while reading the file.",
            level: "error",
          },
        ])
      ),
    [dispatch]
  );
  const onDrop = useCallback(
    (acceptedFiles: Array<File>) => {
      acceptedFiles.forEach((file: File) => {
        const reader = new FileReader();
        reader.onabort = onAbort;
        reader.onerror = onError;
        reader.onload = () => {
          fileUploadCallback(reader.result);
        };
        reader.readAsText(file);
      });
    },
    [fileUploadCallback, onAbort, onError]
  );

  const { getRootProps, getInputProps, isFocused, isDragAccept, isDragReject } =
    useDropzone({ onDrop, accept: mimeTypes, maxFiles: 1, disabled });

  return (
    <div className="container">
      <Container
        {...getRootProps({ isFocused, isDragAccept, isDragReject })}
        aria-label={label ?? "dropzone"}
      >
        <input {...getInputProps()} />
        Drag and drop a file here, or click to select a file.
      </Container>
    </div>
  );
}
