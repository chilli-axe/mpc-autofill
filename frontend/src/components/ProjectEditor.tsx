/**
 * This component represents the complete MPC Autofill project editor, ready to
 * drop into a page (as the only component). Must be wrapped with a Redux provider.
 */

import React, { useEffect } from "react";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { RibbonHeight } from "@/common/constants";
import { useAppSelector } from "@/common/types";
import { NoBackendDefault } from "@/components/NoBackendDefault";
import { OverflowCol } from "@/components/OverflowCol";
import { Ribbon } from "@/components/Ribbon";
import { SelectedImagesRibbon } from "@/features/bulkManagement/SelectedImagesRibbon";
import { CardGrid } from "@/features/card/CardGrid";
import { CommonCardback } from "@/features/card/CommonCardback";
import { Export } from "@/features/export/Export";
import { FinishedMyProject } from "@/features/export/FinishedMyProjectModal";
import { FinishSettings } from "@/features/finishSettings/FinishSettings";
import { Import } from "@/features/import/Import";
import { SearchSettings } from "@/features/searchSettings/SearchSettings";
import { Status } from "@/features/status/Status";
import { useBackendConfigured } from "@/store/slices/backendSlice";
import {
  selectIsProjectEmpty,
  selectProjectCardback,
} from "@/store/slices/projectSlice";

function ProjectEditor() {
  // TODO: should we periodically ping the backend to make sure it's still alive?
  //# region queries and hooks

  const backendConfigured = useBackendConfigured();
  const cardback = useAppSelector(selectProjectCardback);
  const isProjectEmpty = useAppSelector(selectIsProjectEmpty);

  //# endregion

  //# region effects

  /**
   * Ask the user for confirmation before they close the page if their project has any cards in it.
   */
  useEffect(() => {
    const handler = (event: BeforeUnloadEvent) => {
      if (!isProjectEmpty) {
        event.preventDefault();
        return false;
      }
    };
    window.addEventListener("beforeunload", handler);
    return () => {
      window.removeEventListener("beforeunload", handler);
    };
  }, [isProjectEmpty]);

  //# endregion

  return (
    <>
      {backendConfigured ? (
        <Row className="g-0">
          <Ribbon className="g-0">
            <SelectedImagesRibbon />
          </Ribbon>
          <OverflowCol
            lg={8}
            md={8}
            sm={6}
            xs={6}
            data-testid="left-panel"
            heightDelta={RibbonHeight}
          >
            <CardGrid />
          </OverflowCol>
          <OverflowCol
            data-testid="right-panel"
            lg={4}
            md={4}
            sm={6}
            xs={6}
            style={{ zIndex: 1 }}
            className="px-2"
            heightDelta={RibbonHeight}
          >
            <Status />
            <Row className="g-0 pt-2">
              <FinishSettings />
            </Row>
            <Row className="g-0 pt-2">
              <SearchSettings />
            </Row>
            <Row className="g-0 pt-2">
              <Col lg={7} md={12} sm={12} xs={12}>
                <Import />
              </Col>
              <Col lg={5} md={12} sm={12} xs={12}>
                <Export />
              </Col>
              <FinishedMyProject />
            </Row>
            <Col className="g-0 pt-2" lg={{ span: 8, offset: 2 }} md={12}>
              <CommonCardback selectedImage={cardback} />
            </Col>
          </OverflowCol>
        </Row>
      ) : (
        <NoBackendDefault />
      )}
    </>
  );
}

export default ProjectEditor;
