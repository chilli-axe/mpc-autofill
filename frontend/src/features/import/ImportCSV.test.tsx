import { screen } from "@testing-library/react";

import { Back, Front } from "@/common/constants";
import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
  cardDocument4,
  cardDocument5,
  cardDocument6,
} from "@/common/test-constants";
import {
  expectCardbackSlotState,
  expectCardGridSlotState,
  expectCardSlotToExist,
  importCSV,
  openImportCSVModal,
  renderWithProviders,
} from "@/common/test-utils";
import ProjectEditor from "@/components/ProjectEditor";
import {
  cardbacksTwoOtherResults,
  cardDocumentsSixResults,
  cardDocumentsThreeResults,
  defaultHandlers,
  searchResultsFourResults,
  searchResultsOneResult,
  searchResultsSixResults,
  searchResultsThreeResults,
  sourceDocumentsOneResult,
  sourceDocumentsThreeResults,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

const preloadedState = {
  project: {
    members: [],
    cardback: cardDocument2.identifier,
    mostRecentlySelectedSlot: null,
  },
};

//# region snapshot tests

test("the html structure of CSV importer", async () => {
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await openImportCSVModal();

  expect(screen.getByTestId("import-csv")).toMatchSnapshot();
});

//# endregion

test("importing one card by CSV into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import a card
  await importCSV(
    `Quantity,Front
    ,my search query`
  );

  // a card slot should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});

test("importing multiple instances of one card by CSV into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import two instances of a card
  await importCSV(
    `Quantity,Front
    2,my search query`
  );

  // two card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});

test("importing one specific card version by CSV into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import a card
  await importCSV(
    `Quantity,Front,Front ID
    ,my search query,${cardDocument3.identifier}`
  );

  // a card slot should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 3);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});

test("importing one card of each type into an empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsThreeResults,
    searchResultsSixResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import one card of each type
  await importCSV(
    `Quantity,Front
    ,query 1\n,t:query 6\n,b:query 5`
  );

  // three card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument6.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument5.name, 1, 1);
  await expectCardGridSlotState(3, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});

test("importing a more complex CSV into an empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsFourResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import a few cards
  await importCSV(
    `Quantity,Front,Front ID,Back,Back ID
    2,my search query,${cardDocument3.identifier},my search query,${cardDocument4.identifier}
    ,my search query`
  );

  // three card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(1, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(2, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument1.name, 1, 4);
  await expectCardGridSlotState(3, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});

test("CSV header has spaces", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import a card
  await importCSV(
    `Quantity, Front , Front ID
    ,my search query,${cardDocument3.identifier}`
  );

  // a card slot should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 3);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});
