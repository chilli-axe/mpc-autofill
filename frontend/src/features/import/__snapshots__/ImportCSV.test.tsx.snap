// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of CSV importer 1`] = `
.c0 {
  max-width: 100%;
  overflow-x: scroll;
}

.c1 {
  table-layout: fixed;
  border-style: solid;
  border-color: #333333;
  border-width: 1px;
}

.c2 {
  width: 20%;
  text-align: center;
}

.c3 {
  width: 20%;
  text-align: center;
  border-style: solid;
  border-color: #333333;
  border-width: 1px;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 20px;
  border-width: 2px;
  border-radius: 2px;
  border-color: #eeeeee;
  border-style: dashed;
  color: #bdbdbd;
  outline: none;
  -webkit-transition: border 0.24s ease-in-out;
  transition: border 0.24s ease-in-out;
  cursor: pointer;
}

<div
  class="modal-dialog modal-dialog-scrollable"
  data-testid="import-csv"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Add Cards — CSV
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <p>
        Upload a CSV file of cards to add to the project. The file must
         
        <b>
          exactly
        </b>
         match the following format:
      </p>
      <div
        class="c0"
      >
        <table
          class="c1 table"
        >
          <thead>
            <tr>
              <th
                class="c2"
                scope="col"
              >
                Quantity
              </th>
              <th
                class="c2"
                scope="col"
              >
                Front
              </th>
              <th
                class="c2"
                scope="col"
              >
                Front ID
              </th>
              <th
                class="c2"
                scope="col"
              >
                Back
              </th>
              <th
                class="c2"
                scope="col"
              >
                Back ID
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
            </tr>
            <tr>
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
            </tr>
            <tr>
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              />
            </tr>
          </tbody>
        </table>
      </div>
      Where the columns follow these rules:
      <ul>
        <li>
          <b>
            Quantity
          </b>
          : The quantity to include of this row. Must be greater than 0. Can be blank — if blank, a quantity of 1 will be assumed.
        </li>
        <li>
          <b>
            Front
          </b>
          : Search query for card front.
           
          <b>
            Cannot be blank.
          </b>
        </li>
        <li>
          <b>
            Front ID
          </b>
          : If this image is in the front search results, it will be pre-selected. Can be blank.
        </li>
        <li>
          <b>
            Back
          </b>
          : Search query for card back. Can be blank.
        </li>
        <li>
          <b>
            Back ID
          </b>
          : If this image is in the back search results, it will be pre-selected. Can be blank.
        </li>
      </ul>
      For example:
      <div
        class="c0"
      >
        <table
          class="c1 table"
        >
          <thead>
            <tr>
              <th
                class="c2"
                scope="col"
              >
                Quantity
              </th>
              <th
                class="c2"
                scope="col"
              >
                Front
              </th>
              <th
                class="c2"
                scope="col"
              >
                Front ID
              </th>
              <th
                class="c2"
                scope="col"
              >
                Back
              </th>
              <th
                class="c2"
                scope="col"
              >
                Back ID
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="c3"
              >
                <code>
                  2
                </code>
              </td>
              <td
                class="c3"
              >
                <code>
                  island
                </code>
              </td>
              <td
                class="c3"
              >
                <code>
                  1HsvTYs...
                </code>
              </td>
              <td
                class="c3"
              >
                <code>
                  forest
                </code>
              </td>
              <td
                class="c3"
              />
            </tr>
            <tr>
              <td
                class="c3"
              >
                <code>
                  3
                </code>
              </td>
              <td
                class="c3"
              >
                <code>
                  t:goblin
                </code>
              </td>
              <td
                class="c3"
              />
              <td
                class="c3"
              />
              <td
                class="c3"
              >
                <code>
                  1JtXL6C...
                </code>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <hr />
      <div
        class="container"
      >
        <div
          aria-label="import-csv"
          class="c4"
          role="presentation"
          tabindex="0"
        >
          <input
            accept="text/csv,.csv"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          Drag and drop a file here, or click to select a file.
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
